### PR TITLE
Fixes Binding to non-null IEnumerable doesn't work #36390

### PR DESCRIPTION
**Pull request by SteveDunn**
Fixes Binding to non-null IEnumerable doesn't work #36390

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs [hunk 1]
```diff
@@ -220,6 +223,26 @@ public void CanBindConfigurationKeyNameAttributes()
             Assert.Equal("Yo", options.NamedProperty);
         }
 
+        [Fact]
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:226]: do we want to also add a test for null IEnumerable input or that test already exists?

want to make sure we are covering related use cases in test.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs [hunk 2]
```diff
@@ -220,6 +223,26 @@ public void CanBindConfigurationKeyNameAttributes()
...
+        public void CanBindInstantiatedCovariantIEnumerable()
+        {
+            var dic = new Dictionary<string, string>
+            {
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:230]: what happens with input

```c#
            {
                {"InstantiatedCovariantIEnumerable:1", "Yo1"},
                {"InstantiatedCovariantIEnumerable:2", "Yo2"},
            };
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 3]
```diff
@@ -498,6 +500,40 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
...
+                {
+                    object? item = BindInstance(
+                        type: elementType,
+                        instance: null,
+                        config: children[i],
+                        options: options);
+                    if (item != null)
+                    {
+                        singleItemArray[0] = item;
+                        addMethod!.Invoke(source, singleItemArray);
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:526]: We don't want to mutate the source. Please create and return a new list containing source elements and the children elements.

Refer to how similarly with `BindArray` we create a new array and return that one.

**Review by maryamariyan — CHANGES_REQUESTED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs [hunk 4]
```diff
@@ -348,35 +349,6 @@ public void AlreadyInitializedListBinding()
             Assert.Equal("valx", list[4]);
         }
 
-        [Fact]
```
Comment (SteveDunn) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs:351]: Not sure how this was removed. I've re-added it.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs (hunk 1 above)
Comment (SteveDunn) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs:229]: Hi @maryamariyan . Do I need to do anything else for this PR? Do I 'Resolve conversation ' or so you do it?

**Comment by maryamariyan** [comment]
> Hi @maryamariyan . Do I need to do anything else for this PR? Do I 'Resolve conversation ' or so you do it?

@SteveDunn the ones you addressed you could mark as resolved.

Also note, the blank line is still in the diff, you'd want to reconfirm if the commit/changes you wanted is showing up in the PR through the diff tab.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 5]
```diff
@@ -389,6 +390,7 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
             return instance;
         }
 
+
```
[the remark this answers is not in the document]
Comment (SteveDunn) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:403]: Not sure what's gone on here. I'll check tomorrow.

[the remark this answers is not in the document]
Comment (SteveDunn) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:403]: This looks like it's been resolved. At least I can't see a blank line in my branch...

**Review by maryamariyan — APPROVED** [review]
Overall LGTM, but there's some nit changes to be addressed. I think your branch is not in sync with what's present in the PR diff

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:403]:
```suggestion

```

**Review by maryamariyan — APPROVED** [review]

**Review by tarekgh — APPROVED** [review]

**Comment by SteveDunn** [comment]
> Overall LGTM, but there's some nit changes to be addressed. I think your branch is not in sync with what's present in the PR diff

@maryamariyan - should all be OK now.

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
Thanks @SteveDunn!

**Pull request merged**